src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -1396,35 +1399,65 @@ private async Task WriteStreamPropertiesAsync(SelectExpandNode selectExpandNode,
         {
             ISet<IEdmNavigationProperty> navigationProperties = selectExpandNode.SelectedNavigationProperties;
 
-            if (navigationProperties != null)
+            if (navigationProperties == null)
             {
-                IEnumerable<string> changedProperties = null;
-                if (null != resourceContext.EdmObject && resourceContext.EdmObject is IDelta changedObject)
+                yield break;
+            }
+
+            IEnumerable<string> changedProperties = null;
+
+            if (resourceContext.EdmObject != null && resourceContext.EdmObject is IDelta changedObject)
+            {
+                changedProperties = changedObject.GetChangedPropertyNames();
+
+                foreach (IEdmNavigationProperty navigationProperty in navigationProperties)
+                {
+                    if (changedProperties == null || changedProperties.Contains(navigationProperty.Name))
+                    {
+                        yield return new KeyValuePair<IEdmNavigationProperty, Type>(navigationProperty, typeof(IEdmChangedObject));
+                    }
+                }
+            }
+            else if (resourceContext.ResourceInstance != null && resourceContext.ResourceInstance is IDelta deltaObject)
+            {
+                changedProperties = deltaObject.GetChangedPropertyNames();
+                dynamic delta = deltaObject;
+
+                foreach (IEdmNavigationProperty navigationProperty in navigationProperties)
                 {
-                    changedProperties = changedObject.GetChangedPropertyNames();
+                    Object obj = null;
 
-                    foreach (IEdmNavigationProperty navigationProperty in navigationProperties)
+                    if (changedProperties == null || changedProperties.Contains(navigationProperty.Name) && delta.DeltaNestedResources.TryGetValue(navigationProperty.Name, out obj))
```

Comment:
Add some brackets in the conditional expressions to make it clear whether `changedProperties == null || changedProperties.Contains(navigationProperty.Name)` or `changedProperties.Contains(navigationProperty.Name) && delta.DeltaNestedResources.TryGetValue(navigationProperty.Name, out obj)` is evaluated first

Reply:
I think Sreejith left that code that way intentionally. https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#operator-precedence

Reply:
@KenitoInc RE: https://github.com/OData/WebApi/pull/2674/files#r979664329
Operator precedence tables are one thing and providing clarity to the naked eye is another. It's for the same reason why I wouldn't feel comfortable with this: `18 / 2 * 3`. The compiler will have no problem interpreting that but a fair number of developers would have to refer to the operator precedence tables to interpret it accurately.

Reply:
Refactored code. So this now is not an issue